File: apps/demo-domain/rules-service/src/workitem-processor.ts
import { ProcessedWorkItem } from '../../../../packages/proto/generated/ts/pipeline/v1/pipeline'

export interface EnrichedWorkItem {
  id: string
  eigenvalues: number[]
  score: number
  processing_time_ms?: number
}

/**
 * Processes a ProcessedWorkItem from parse-service
 * Simulates Python numpy operations (feature engineering)
 */
export function processWorkItem(processedItem: any): EnrichedWorkItem {
  const start = Date.now()

  // Extract normalized vectors
  const vectors: number[][] =
    processedItem.normalized_vectors?.map((v: any) => v.values || []) || []

  // Compute statistics (mean, std) from all vector values
  const allValues: number[] = []
  for (const vec of vectors) {
    allValues.push(...vec)
  }

  const mean = allValues.length > 0 ? allValues.reduce((s, v) => s + v, 0) / allValues.length : 0
  const variance =
    allValues.length > 0
      ? allValues.reduce((s, v) => s + Math.pow(v - mean, 2), 0) / allValues.length
      : 0
  const std = Math.sqrt(variance)

  // Use first 5 values as "eigenvalues" (simplified)
  const eigenvalues = allValues.slice(0, 5)

  // Matrix score: sum all elements from transposed matrix
  const matrix = processedItem.transposed_matrix
  let matrixScore = 0.0
  if (matrix?.rows) {
    for (const row of matrix.rows) {
      if (row.values) {
        matrixScore += row.values.reduce((s: number, v: number) => s + v, 0)
      }
    }
  }

  // CPU-intensive iterations
  const iterations = 500
  let enrichedScore = 0.0
  const checksum = processedItem.checksum || 0
  for (let i = 0; i < iterations; i++) {
    enrichedScore += (checksum + matrixScore + mean) * (i + 1) * 0.001
  }

  const processingTime = Date.now() - start

  return {
    id: processedItem.id,
    eigenvalues,
    score: enrichedScore,
    processing_time_ms: processingTime,
  }
}

File: apps/demo-domain/rules-service/src/workitem_processor.py
"""WorkItem processing for rules-service"""
import json
import time
from typing import Dict, List, Any


def process_work_item(processed_item: Dict[str, Any]) -> Dict[str, Any]:
    """
    Processes a ProcessedWorkItem from parse service.
    Applies feature engineering and CPU-intensive operations.
    """
    start = time.perf_counter()

    item_id = processed_item.get('id', 'unknown')
    normalized_vectors = processed_item.get('normalizedVectors', [])
    transposed_matrix = processed_item.get('transposedMatrix', {})
    checksum = processed_item.get('checksum', 0.0)

    # Feature engineering: compute statistics
    vector_stats = compute_vector_statistics(normalized_vectors)

    # Matrix operations: compute eigenvalues (simplified)
    matrix_score = compute_matrix_score(transposed_matrix)

    # CPU-intensive iterations
    iterations = 500  # Fixed for now
    enriched_score = 0.0
    for i in range(iterations):
        enriched_score += (checksum + matrix_score + vector_stats['mean']) * (i + 1) * 0.001

    processing_time = (time.perf_counter() - start) * 1000

    return {
        'id': item_id,
        'eigenvalues': vector_stats['values'][:5],  # First 5 values as "eigenvalues"
        'score': enriched_score,
        'processing_time_ms': processing_time
    }


def compute_vector_statistics(vectors: List[Dict[str, List[float]]]) -> Dict[str, Any]:
    """Computes statistics over normalized vectors"""
    all_values = []
    for vec in vectors:
        all_values.extend(vec.get('values', []))

    if not all_values:
        return {'mean': 0.0, 'std': 0.0, 'values': []}

    mean = sum(all_values) / len(all_values)
    variance = sum((x - mean) ** 2 for x in all_values) / len(all_values)
    std = variance ** 0.5

    return {
        'mean': mean,
        'std': std,
        'values': all_values
    }


def compute_matrix_score(matrix: Dict[str, List[Dict[str, List[float]]]]) -> float:
    """Computes a score from matrix (sum of all elements)"""
    rows = matrix.get('rows', [])
    total = 0.0
    for row in rows:
        values = row.get('values', [])
        total += sum(values)
    return total
